File: python/step/download_files.py
import asyncio
import os

from utils.download_files import download_file


async def download_files(m3u_file: str, urls: list[str]):
    """
    Downloads all URLs in the application's state.
    """
    m3u_file_name = os.path.splitext(os.path.basename(m3u_file))[0]

    async def run_downloader():
        await asyncio.gather(*[download_file(url, m3u_file_name) for url in urls])

    print(f"Starting download of {len(urls)} files...")
    await run_downloader()
    print("All files downloaded.")

File: ruby/lib/download_files.rb
# frozen_string_literal: true

require 'async'
require_relative 'download_file'

module DownloadFiles
  def self.download(m3u_file, urls)
    m3u_file_name = File.basename(m3u_file, '.*')

    puts "Starting download of #{urls.length} files..."

    Async do
      barrier = Async::Barrier.new
      urls.each do |url|
        barrier.async do
          DownloadFile.download(url, m3u_file_name)
        end
      end
      barrier.wait
    end

    puts 'All files downloaded.'
  end
end
